adding tutorial (#46)
tgm-team/tgm: docs/tutorials/dgraph_tutorial.md

# Constructing and Accessing Properties in DGraph

> This tutorial will show you how to construct a `DGraph` object in `tgm` and accessing its properties.

Reviewer: ```suggestion
This tutorial shows how to construct a `DGraph` object in `tgm` and explore its properties..
```

> The `DGraph` object can be found in `tgm/graph.py`

Reviewer: ```suggestion
The `DGraph` class is defined in [`tgm/graph.py`](https://github.com/tgm-team/tgm/blob/main/tgm/graph.py).
```

> Let's first learn how to make a `DGraph`by defining the graph ourselves.

Reviewer: Maybe we should start the tutorial with the simplest and most common use case: move TGB subfolder here, since defining with custom tensors is most advanced construction use case

Author: yes, will discuss `from_tgb` here

Start by defining the temporal edges, the `edge_index` has shape $\[num_edge_events, 2\]$

The `edge_timestamps` tensor specifies the timestamps associated with each edge, must has shape $\[num_edge_events\]$

The `edge_feats` specifies the features associated with each edge, must has shape $\[num_edge_events, D_edge\]$ where $D_edge$ is the feature dimension of edge. Edge features are also optional

```python
import torch

edge_index = torch.LongTensor([[2, 2], [2, 4], [1, 8]])
edge_timestamps = torch.LongTensor([1, 5, 20])
edge_feats = torch.rand(3, 5)
```

## Construct `DGraph` from raw tensors

> To construct a `DGraph` object, we use `DGraph.from_raw` method, seen in [`tgm/graph.py`](https://github.com/tgm-team/tgm/blob/main/tgm/graph.py). In TGM, we support both edge events and node events at the same time. The core assumption is that a temporal graph must have edge events and with optional node events as well. You can also easily specify the time granularity of your data with a single string, for example we will specify a second wise granularity, as 's'. Similarly, we can also specify the device.

Reviewer: We might need to properly introduce the notion of time delta in order for the user to understand how to interpret and use this argument.

    node_df=pd.DataFrame(node_dict),
    node_id_col='node',
    node_time_col='t',
    dynamic_node_feats_col='node_feat',
    time_delta='s',
    device='cpu',
)
```

## Construct `DGraph` from CSVs

We can also construct `DGraph` directly from csv files by using `DGraph.from_csv` method. see [`tgm/graph.py`](https://github.com/tgm-team/tgm/blob/main/tgm/graph.py) for details.

## Properties of `DGraph`

`DGraph` objects act as a view on the underlying graph data, allowing the user to access various properties as well as slicing and other operations. Additional properties are seen in the `tgm/graph.py`

```python
print('=== Graph Properties ===')
print(f'start time : {our_dgraph.start_time}')
print(f'end time : {our_dgraph.end_time}')
print(f'number of nodes : {our_dgraph.num_nodes}')
print(f'number of edge events : {our_dgraph.num_edges}')
print(f'number of timestamps : {our_dgraph.num_timestamps}')
print(f'number of edge and node events : {our_dgraph.num_events}')
print(f'edge feature dim : {our_dgraph.edge_feats_dim}')
print(f'static node feature dim : {our_dgraph.static_node_feats_dim}')
print(f'dynamic node feature dim : {our_dgraph.dynamic_node_feats_dim}')

> print('======================')

Reviewer: Would be good to clarify how we compute / handle these, especially on subgraphs (e.g. start time is None on empty slice). Number of nodes is based on the max  id + 1.

```

## Construct `DGraph` from TGB Datasets

We support the construction of `DGraph` from TGB link property prediction and node property prediction datasets. Currently, TKG and THG datasets from TGB is not supported. specify the time granularity of the dataset with `TimeDeltaDG`. Note that time granularity related functions are in alpha, and will be still under change

`r` means ordered time granularity which means the underlining timestamps tell us the ordering of edges and we do not use it for time conversion. (other time granularities will allow)

```python
from tgm.timedelta import TimeDeltaDG

train_dg = DGraph('tgbl-wiki', time_delta=TimeDeltaDG('r'), split='train')
```

> ## `DGDataLoader` and `DGHook`

Reviewer: Should this be part of DGraph tutorial? Seems more advanced than just learning how to construct and query properties on the graph.

Author: for alpha user and complete basic workflow this is the foundation, I would in future PR create separate tutorial for this. Currently this tutorial includes the minimum you need to complete workflow.
